src/workflow/confirm.c

```diff
@@ -18,48 +18,131 @@
 #include "blocking.h"
 #include "hardfault.h"
 
+#include <hardfault.h>
+#include <ui/components/confirm.h>
 #include <ui/screen_stack.h>
+#include <ui/workflow_stack.h>
+#include <util.h>
 
 #include <stddef.h>
+#include <stdlib.h>
 
-static bool _result = false;
+typedef struct {
+    bool result;
+    bool done;
+    confirm_params_t params;
+    void (*callback)(bool, void*);
+    void* callback_param;
+    char* title;
+    char* body;
+} data_t;
 
-static void _confirm(component_t* component)
+static void _confirm(void* param)
 {
-    (void)component;
-    _result = true;
-    workflow_blocking_unblock();
+    workflow_t* self = (workflow_t*)param;
+    data_t* data = (data_t*)self->data;
+    data->result = true;
+    data->done = true;
 }
 
-static void _reject(component_t* component)
+static void _reject(void* param)
 {
-    (void)component;
-    _result = false;
-    workflow_blocking_unblock();
+    workflow_t* self = (workflow_t*)param;
+    data_t* data = (data_t*)self->data;
+    data->result = false;
+    data->done = true;
+}
+
+/**
+ * Checks if the user has confirmed the choice.
+ */
+static void _workflow_confirm_spin(workflow_t* self)
+{
+    data_t* data = (data_t*)self->data;
+    if (data->done) {
+        /* Publish our result. */
+        data->callback(data->result, data->callback_param);
+        /* Time to go, goodbye. */
+        workflow_stack_stop_workflow();
+    }
 }
 
-static bool _have_result = false;
+/**
+ * Starts this workflow.
+ */
+static void _workflow_confirm_init(workflow_t* self)
+{
+    data_t* data = (data_t*)self->data;
+    component_t* comp;
+    comp = confirm_create(
+        &data->params, _confirm, self, data->params.accept_only ? NULL : _reject, self);
+    ui_screen_stack_push(comp);
+}
 
-static void _confirm_async(component_t* component)
+/**
+ * Destroys this workflow.
+ */
+static void _workflow_confirm_cleanup(workflow_t* self)
 {
-    (void)component;
-    _result = true;
-    _have_result = true;
     ui_screen_stack_pop();
+    ui_screen_stack_cleanup();
+    util_zero(self->data, sizeof(data_t));
+    free(self->data);
+    util_zero(self, sizeof(*self));
+    free(self);
+}
+
+workflow_t* workflow_confirm(
+    const confirm_params_t* params,
+    void (*callback)(bool, void*),
+    void* callback_param)
+{
+    workflow_t* result = workflow_allocate(
+        _workflow_confirm_init, _workflow_confirm_cleanup, _workflow_confirm_spin);
+    data_t* data = malloc(sizeof(*data));
+    if (!data) {
```

Comment:
it's a good idea to memset 0 all malloced structs to avoid accidental undefined data (e.g. when the struct is extended).

then you could also skip the ` = NULL`, `=false` parts.

Reply:
I'm thinking of just moving this inside the `workflow_allocate` function. Every workflow is going to need some private data anyway :|

Reply:
workflow_allocate doesn't know what the data struct is or how big it is.

Reply:
We can just pass `sizeof(data_t)` in `workflow_allocate` as a parameter. The `data` field is `void *` anyway.

Reply:
Reorganized this so that `workflow_allocate` handles private data storage automatically - fcca89b

Reply:
Also 39b48f4 - Allocated data is initialized to zeros.